### PR TITLE
Improve display of CustomDebugInformation records

### LINKED ISSUE #83: mdv /out option appears to not truncate content before overwriting
Repro steps:

1. Run mdv on some dll with an /out argument to some file path
2. Run it again to the same path but with something that will give simpler output

Expected behavior: Content of the old output file is removed and replaced with the new output
Actual behavior: Trailing content of the old output remains as garbage in the new output file

### LINKED ISSUE #97: Improve display of known CDIs
Some CDIs have customized printing while others only print the default hex dump.

Custom:
```
1: 0x00000001 (Module)     Source Link (#1)                      7B-22-64-6F-63-75-6D-65-6E-74-73-22-3A-7B-22-44-3A-5C-5C-73-6F-75-72-63-65-6C-69-6E-6B-5C-5C-2A-... (#1)
{"documents":{"D:\\sourcelink\\*":"https://raw.githubusercontent.com/tmat/sourcelink/92f1252a59eb67391c1883a5c39920a385157f1a/*"}}
```

Default:
```
2: 0x30000005 (Document)   Embedded Source (#8)                  DD-07-00-00-A5-94-6D-6F-DA-30-10-C7-DF-23-F1-1D-6E-DD-0B-60-5A-43-B7-97-7D-5C-C5-58-85-36-3A-A9-... (#2af7)
```

The proposal is to decode the CDIs that are using the default hex dump and print content in human readable form.

Also consider if displaying hex dump as well as human readable form is useful (perhaps add an option that controls that behavior).

### PR DESCRIPTION
**Pull request by samaw7**
Changes made for formatting issue in CustomDebugInformation #2 compilation options was on the same line as #1. Also added data for Embedded sources, it now displays the source.

Fixes #97
Fixes #83

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/metadata-tools?pullRequest=101) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/metadata-tools?pullRequest=101) before we can accept your contribution.<br/><br/>:x: samuelawuah [sign now](https://cla.dotnetfoundation.org/dotnet/metadata-tools?pullRequest=101)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/metadata-tools?pullRequest=101) it.</sub>

**Comment by tmat** [comment]
@samuelawuah Will take a look!

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 1]
```diff
@@ -2054,7 +2055,7 @@ public void WriteCustomDebugInformation()
             {
                 var entry = _reader.GetCustomDebugInformation(handle);
 
-                table.AddRowWithDetails(
+                table.AddRowWithDetails(
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2058]: Nit: extra space at the end of the line

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 2]
```diff
@@ -2111,12 +2112,16 @@ public string TryDecodeCustomDebugInformation(CustomDebugInformation entry)
             {
                 return VisualizeCompilationOptions(blobReader);
             }
+            if(kind ==  PortableCustomDebugInfoKinds.EmbeddedSource)
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2115]: Style: Add new line before `if` and format this line with `Ctrl+K+F`.

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 3]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
             builder.AppendLine("}");
             return builder.ToString();
         }
+        private static string VisualizeEmbeddedSource(BlobReader reader)
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2223]: Style: add empty line before `private`

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 4]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+        {
+            var utf8 = new UTF8Encoding();
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2225]: You can use predefined instance `Encoding.UTF8`

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 5]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+            var builder = new StringBuilder();
+            builder.AppendLine(">>>");
+            int format = -1;
+            try { format = reader.ReadInt32(); } catch { };
+            byte[] bytes= null;
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2230]: Style: space before `=`

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 6]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+       
+            }
+            if (format < 0 || bytes == null)
+            {
+                builder.AppendLine(BadMetadataStr);
+            }
+            else
+            {
+                builder.AppendLine(utf8.GetString(bytes));
+                builder.AppendLine("### End of Embedded Source");
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2258]: I'd use `<<<` instead of `###` to be consistent with `>>>` above.

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 7]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+                    deflate.CopyTo(uncompressedStream);
+                    bytes = uncompressedStream.ToArray();
+                }
+                catch 
+                {
+                    bytes = null;
+                };
+       
+            }
+            if (format < 0 || bytes == null)
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2251]: You can simplify the conditions using `else if` like so:

```
if (format < 0 || bytes == null)
{
  ...
}
else if (format > 0) 
{
}
else
{
}
```

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 8]
```diff
@@ -2215,6 +2220,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+                    using var uncompressedStream = new MemoryStream();
+                    using var deflate = new DeflateStream(compressedStream, CompressionMode.Decompress);
+                    deflate.CopyTo(uncompressedStream);
+                    bytes = uncompressedStream.ToArray();
+                }
+                catch 
+                {
+                    bytes = null;
+                };
+
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2249]: Style: extra blank line

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs (hunk 7 above)
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2251]: Style: add empty line before `if`

**Comment by tmat** [comment]
Looks good, just a few minor stylistic details. Once you correct them, create another commit and push. GitHub will automatically update the PR.

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 9]
```diff
@@ -2216,6 +2222,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+            try { format = reader.ReadInt32(); } catch { };
+            byte[] bytes = null;
+            try { bytes = reader.ReadBytes(reader.RemainingBytes); } catch { };
+                
+           
+            if (format > 0 && bytes != null)
+            {
+                try
+                {
+
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2239]: One more extra empty line

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 10]
```diff
@@ -2216,6 +2222,45 @@ private static string VisualizeCompilationOptions(BlobReader reader)
...
+        private static string VisualizeEmbeddedSource(BlobReader reader)
+        {
+            var builder = new StringBuilder();
+            builder.AppendLine(">>>");
+            int format = -1;
+            try { format = reader.ReadInt32(); } catch { };
+            byte[] bytes = null;
+            try { bytes = reader.ReadBytes(reader.RemainingBytes); } catch { };
+                
+
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2234]: extra empty line

**Comment by RikkiGibson** [comment]
It looks like empty file 'pdb2.txt' may not have been intended to be included in the PR.

File: src/mdv/Mdv.cs [hunk 11]
```diff
@@ -40,7 +40,7 @@ public GenerationData(IDisposable memoryOwner, MetadataReader metadataReader, PE
         public Mdv(Arguments arguments)
         {
             _arguments = arguments;
-            _writer = (arguments.OutputPath != null) ? new StreamWriter(File.OpenWrite(arguments.OutputPath), Encoding.UTF8) : Console.Out;
+            _writer = (arguments.OutputPath != null) ? new StreamWriter(File.Create(arguments.OutputPath), Encoding.UTF8) : Console.Out;
```
Comment (RikkiGibson) [inline src/mdv/Mdv.cs:43]: Is this change related to #83?

**Comment by RikkiGibson** [comment]
It would be nice if there were tests added for these changes, but it looks like there are not many tests in this project to use as a starting point. Perhaps in a later PR.

File: src/mdv/Mdv.cs (hunk 11 above)
Comment (tmat) [reply src/mdv/Mdv.cs:43]: Yes.

**Comment by tmat** [comment]
@RikkiGibson We don't currently have direct tests for the MetadataVisualizer. It's mostly tested thru Roslyn. In future we might want to add some tests.

File: src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs [hunk 12]
```diff
@@ -2112,11 +2113,16 @@ public string TryDecodeCustomDebugInformation(CustomDebugInformation entry)
                 return VisualizeCompilationOptions(blobReader);
             }
 
+            if (kind == PortableCustomDebugInfoKinds.EmbeddedSource)
```
Comment (tmat) [inline src/Microsoft.Metadata.Visualizer/MetadataVisualizer.cs:2116]: Check `MetadataVisualizerOptions.EmbeddedSources` flag

**Review by tmat — APPROVED** [review]
